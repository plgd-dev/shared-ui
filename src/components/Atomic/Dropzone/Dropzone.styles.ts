import { css } from '@emotion/react'
import { ThemeType, getThemeColor } from '../_theme'

export const dropzoneContainer = (theme: ThemeType) => css`
    padding: 32px;
    border: 1px dashed ${getThemeColor(theme, `Dropzone.container.borderColor`)};
    background: ${getThemeColor(theme, `Dropzone.container.background`)};
    border-radius: 8px;
    cursor: pointer;
`

export const smallPadding = css`
    padding: 16px;
`

export const placeholder = (theme: ThemeType) => css`
    display: flex;
    flex-direction: column;
    align-items: center;
    color: ${getThemeColor(theme, `Dropzone.placeholder.color`)};
`

export const placeholderText = (theme: ThemeType) => css`
    color: ${getThemeColor(theme, `Dropzone.placeholder.text.color`)};
    font-weight: 400;
    font-size: 14px;
    line-height: 22px;
    margin: 12px 0 0 0;

    span {
        color: ${getThemeColor(theme, `Dropzone.placeholder.highlight.color`)};
        text-decoration: underline;
        cursor: pointer;
    }
`

export const placeholderDescription = (theme: ThemeType) => css`
    color: ${getThemeColor(theme, `Dropzone.placeholder.description.color`)};
    font-weight: 400;
    font-size: 12px;
    line-height: 18px;
    margin: 4px 0 0 0;
`

export const imageRow = css`
    display: flex;
`

export const imageWrapper = css`
    flex: 0 0 40px;
`

export const contentWrapper = css`
    flex: 1 1 auto;
`

export const image = css`
    display: block;
    width: 100%;
    height: auto;
`

export const fileLine = css`
    display: flex;
    justify-content: space-between;
    flex: 1 1 auto;
    padding-left: 12px;
`

export const fileName = (theme: ThemeType) => css`
    color: ${getThemeColor(theme, `Dropzone.file.color`)};
    font-weight: 400;
    font-size: 14px;
    line-height: 22px;
`

export const closeIcon = (theme: ThemeType) => css`
    cursor: pointer;
    color: ${getThemeColor(theme, `Dropzone.file.close.color`)};
`

export const fileSize = (theme: ThemeType) => css`
    color: ${getThemeColor(theme, `Dropzone.file.size.color`)};
    font-weight: 400;
    font-size: 12px;
    line-height: 18px;
`

export const progressBar = (theme: ThemeType) => css`
    margin-top: 8px;
    background: ${getThemeColor(theme, `Dropzone.file.progressBar.background`)};
    height: 4px;
    border-radius: 52px;
`

export const fileRejections = css`
    padding-top: 8px;
`

export const rejectedFile = (theme: ThemeType) => css`
    color: ${getThemeColor(theme, `Dropzone.rejectedFile.color`)};
`
