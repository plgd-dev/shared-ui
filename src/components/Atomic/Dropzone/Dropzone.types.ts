import { ReactNode } from 'react'
import { Accept, FileError } from 'react-dropzone'

export type CustomFileRenderType = {
    format: string
    icon: string
}

export type Props = {
    accept?: Accept
    customFileRenders?: CustomFileRenderType[]
    description?: ReactNode
    disabled?: boolean
    maxFiles?: number
    maxSize?: number
    title: string
    validator?: <T extends File>(file: T) => FileError | FileError[] | null
}
