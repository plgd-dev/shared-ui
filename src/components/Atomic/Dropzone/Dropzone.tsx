import { cloneElement, FC, Fragment, ReactElement, SyntheticEvent, useCallback, useEffect, useMemo, useState } from 'react'
import { useDropzone } from 'react-dropzone'
import isFunction from 'lodash/isFunction'

import { defaultProps, Props } from './Dropzone.types'
import * as styles from './Dropzone.styles'
import { convertSize, Icon, IconClose, IconFileUpload } from '../Icon'

const Dropzone: FC<Props> = (props) => {
    const { accept, customFileRenders, description, disabled, maxFiles, maxSize, onFileDrop, onFilesDrop, renderThumbs, smallPadding, title, validator } = {
        ...defaultProps,
        ...props,
    }
    const [files, setFiles] = useState<any>([])

    const onDrop = useCallback(
        (acceptedFiles: any) => {
            setFiles([
                ...files,
                ...acceptedFiles.map((file: any) =>
                    Object.assign(file, {
                        preview: URL.createObjectURL(file),
                    })
                ),
            ])

            const droppedProm = acceptedFiles.map(
                (file: any) =>
                    new Promise(function (resolve, reject) {
                        const reader = new FileReader()

                        reader.onabort = () => {
                            console.log('file reading was aborted')
                            reject('file reading was aborted')
                        }
                        reader.onerror = () => {
                            console.log('file reading has failed')
                            reject('file reading has failed')
                        }
                        reader.onload = () => {
                            isFunction(onFileDrop) && onFileDrop(reader.result)
                            resolve(reader.result)
                        }

                        reader.readAsText(file)
                    })
            )

            Promise.all(droppedProm).then((loadedFiles) => {
                isFunction(onFilesDrop) && onFilesDrop(loadedFiles)
            })
        },
        [files, onFileDrop, onFilesDrop]
    )

    const { getRootProps, getInputProps, fileRejections } = useDropzone({
        accept,
        disabled,
        maxFiles,
        maxSize,
        onDrop,
        validator,
    })

    const renderFile = useCallback(
        (file: any) => {
            const nameArray = file.name.split('.')
            const format = nameArray[nameArray.length - 1]
            const index = customFileRenders?.findIndex((item) => item.format === format.toLowerCase())

            if (customFileRenders && index !== undefined && index >= 0) {
                return typeof customFileRenders[index].icon === 'string' ? (
                    <Icon icon={customFileRenders[index].icon as string} size={40} />
                ) : (
                    cloneElement(customFileRenders[index].icon as ReactElement, { ...convertSize(40) })
                )
            }

            return (
                <img
                    alt={file.name}
                    css={styles.image}
                    onLoad={() => {
                        URL.revokeObjectURL(file.preview)
                    }}
                    src={file.preview}
                />
            )
        },
        [customFileRenders]
    )

    const removeFile = (file: any, e: SyntheticEvent) => {
        e.stopPropagation()
        const newFiles = [...files]
        newFiles.splice(newFiles.indexOf(file), 1)
        setFiles(newFiles)
    }

    function bytesFormatter(num: number) {
        let i = 0
        for (i = 0; num >= 1024 && i < 4; i++) {
            num /= 1024
        }
        return [`${num.toFixed(2)} `, [' Bytes', ' kB', 'MB', 'GB', 'TB'][i]]
    }

    const thumbs = useMemo(
        () =>
            files.map((file: any, key: number) => (
                <Fragment key={key}>
                    <div css={styles.imageRow} key={file.name}>
                        <div css={styles.imageWrapper}>{renderFile(file)}</div>

                        <div css={styles.contentWrapper}>
                            <div css={styles.fileLine}>
                                <div css={styles.fileName}>{file.name}</div>
                                <IconClose css={styles.closeIcon} onClick={(e) => removeFile(file, e)} />
                            </div>
                            <div css={styles.fileLine}>
                                <div css={styles.fileSize}>{bytesFormatter(file.size)}</div>
                                <div css={styles.fileSize}>100%</div>
                            </div>
                        </div>
                    </div>
                    <div css={styles.progressBar}></div>
                </Fragment>
            )),
        // eslint-disable-next-line react-hooks/exhaustive-deps
        [files, renderFile]
    )

    useEffect(() => {
        // Make sure to revoke the data uris to avoid memory leaks, will run on unmount
        return () => files.forEach((file: any) => URL.revokeObjectURL(file.preview))
    }, [files])

    return (
        <div>
            <div {...getRootProps({ className: 'dropzone' })} css={[styles.dropzoneContainer, smallPadding && styles.smallPadding]}>
                <input {...getInputProps()} />
                {thumbs.length !== maxFiles && (
                    <div css={styles.placeholder}>
                        <IconFileUpload {...convertSize(50)} />

                        <p css={styles.placeholderText}>{title}</p>
                        {description && <p css={styles.placeholderDescription}>{description}</p>}
                    </div>
                )}

                {renderThumbs && thumbs && <aside>{thumbs}</aside>}
            </div>
            {fileRejections.length > 0 && (
                <div css={styles.fileRejections}>
                    {fileRejections.map((file, key) => (
                        <div css={styles.rejectedFile} key={key}>
                            {file.file.name}:
                            {file.errors.map((e, j) => (
                                <span key={j}>{e.message}</span>
                            ))}
                        </div>
                    ))}
                </div>
            )}
        </div>
    )
}

Dropzone.displayName = 'Dropzone'

export default Dropzone
