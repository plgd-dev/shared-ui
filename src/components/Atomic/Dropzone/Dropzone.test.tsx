import { render } from '../../../../test/jest-wrapper'
import Dropzone from './Dropzone'

describe('<ConditionalWrapper>', () => {
    it('render correctly - snapshot', () => {
        const { asFragment } = render(
            <div>
                <Dropzone title='Title' />
            </div>
        )

        expect(asFragment()).toMatchSnapshot()
    })
})
