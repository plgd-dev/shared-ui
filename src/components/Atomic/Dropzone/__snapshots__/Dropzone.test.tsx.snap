// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConditionalWrapper> render correctly - snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding: 32px;
  border: 1px dashed #D7D8DA;
  border-radius: 8px;
  cursor: pointer;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #81868C;
}

.emotion-2 {
  color: #252626;
  font-weight: 400;
  font-size: 14px;
  line-height: 22px;
  margin: 12px 0 0 0;
}

.emotion-2 span {
  color: #252626;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

<div>
    <div>
      <div
        class="dropzone emotion-0"
        role="presentation"
        tabindex="0"
      >
        <input
          multiple=""
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        <div
          class="emotion-1"
        >
          <svg
            fill="none"
            height="50"
            viewBox="0 0 50 61"
            width="50"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M.5 6A5.5 5.5 0 0 1 6 .5h23.046A5.5 5.5 0 0 1 32.9 2.077l3.25 3.192 3.705 3.64a5.5 5.5 0 0 1 1.645 3.923V50a5.5 5.5 0 0 1-5.5 5.5H6A5.5 5.5 0 0 1 .5 50z"
              fill="#F6F7F9"
              stroke="#E6E9ED"
            />
            <mask
              height="56"
              id="icon-file-upload_svg__a"
              maskUnits="userSpaceOnUse"
              style="mask-type: alpha;"
              width="42"
              x="0"
              y="0"
            >
              <path
                d="M.5 6A5.5 5.5 0 0 1 6 .5h23.046A5.5 5.5 0 0 1 32.9 2.077l3.25 3.192 3.705 3.64a5.5 5.5 0 0 1 1.645 3.923V50a5.5 5.5 0 0 1-5.5 5.5H6A5.5 5.5 0 0 1 .5 50z"
                fill="#F6F7F9"
                stroke="#D7D8DA"
              />
            </mask>
            <g
              mask="url(#icon-file-upload_svg__a)"
            >
              <path
                d="M30 0h12v12h-9a3 3 0 0 1-3-3z"
                fill="#E6E9ED"
              />
            </g>
            <circle
              cx="34"
              cy="45"
              fill="#2261AE"
              r="16"
            />
            <path
              d="M29.417 50.833a3.751 3.751 0 0 1-1.245-7.29l-.005-.21a4.166 4.166 0 0 1 7.779-2.078 2.5 2.5 0 0 1 3.721 2.98A3.333 3.333 0 0 1 39 50.832zm2.057-5.651a.625.625 0 0 0 .883.883l1.016-1.016v3.492c0 .346.278.625.625.625a.623.623 0 0 0 .625-.625v-3.492l1.015 1.016a.625.625 0 0 0 .883-.883l-2.083-2.084a.62.62 0 0 0-.883 0l-2.083 2.084z"
              fill="#fff"
            />
          </svg>
          <p
            class="emotion-2"
          >
            Title
          </p>
        </div>
        <aside />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
